src/pkg/golang/compiler.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2015-2024 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package golang is an API to the Go compiler.
+package golang
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+type CompilerType int
+
+const (
+	CompilerGo CompilerType = iota
+	CompilerTinygo
+	CompilerUnkown
+)
+
+// cached information re: the compiler
+type Compiler struct {
+	Path          string
+	Identifier    string // e.g. 'tinygo' or 'go'
+	Type          CompilerType
+	Version       string // compiler-tool version, e.g. '0.32.0' for tinygo, go1.22.2 for
+	VersionGo     string // version of go: same as 'Version' for standard go
+	VersionOutput string // output of calling 'tool version'
+	IsInit        bool   // CompilerInit() succeeded
+}
+
+// map the compiler's identifier ("tinygo" or "go") to enum
+func CompilerTypeFromString(name string) CompilerType {
+	val, ok := map[string]CompilerType{
+		"go":     CompilerGo,
+		"tinygo": CompilerTinygo,
+	}[name]
+	if ok {
+		return val
+	}
+	return CompilerUnkown
+}
+
+// WithCompiler sets the compiler for Build() / BuildDir()
+func WithCompiler(p string) Opt {
+	return func(c *Environ) {
+		c.Compiler.Path = p
+		c.Compiler.IsInit = false
+	}
+}
+
+// compilerCmd returns a compiler command to be run in the environment.
+func (c Environ) compilerCmd(gocmd string, args ...string) *exec.Cmd {
+	goBin := c.Compiler.Path
+	if "" == goBin {
+		goBin = filepath.Join(c.GOROOT, "bin", "go")
+	}
+	args = append([]string{gocmd}, args...)
+	cmd := exec.Command(goBin, args...)
+	if c.GBBDEBUG {
+		log.Printf("GBB Go invocation: %s %s %#v", c, goBin, args)
+	}
+	cmd.Dir = c.Dir
+	cmd.Env = append(os.Environ(), c.Env()...)
+	return cmd
+}
+
+// if go-compiler specified, resolve absolute-path from PATH,
+// otherwise return 'nil'.
+func (c *Environ) compilerAbs() error {
+	if c.Compiler.Path != "" {
+		fname, err := exec.LookPath(string(c.Compiler.Path))
+		if err == nil {
+			fname, err = filepath.Abs(fname)
+		}
+		if err != nil {
+			return fmt.Errorf("build: %v", err)
+		}
+		c.Compiler.Path = fname
+	}
+	return nil
+}
+
+// runs compilerCmd("version") and parse/caches output, to environ.Compiler
+func (c *Environ) CompilerInit() error {
+
+	if c.Compiler.IsInit {
+		return nil
+	}
+
+	c.compilerAbs()
+
+	cmd := c.compilerCmd("version")
+	v, err := cmd.CombinedOutput()
```

Comment:
There is no occurance in your code that uses `v` as raw `[]bytes` returned by `cmd.CombinedOutput`. Consider making `v` a string and get rid of all the `string(v)` if you never need `v` itself anyways.